### ISSUE TITLE
[FragmentProgram] Allow defining an ImageFilter

### ISSUE THREAD

**Issue by flar** [flutter/flutter#96823]
Similar to #94235 the recent work to enable the creation of custom Shader objects is very close in functionality to enabling the creation of custom ImageFilter objects. What would be needed is to allow the program to access an image input that will be supplied at render time and then wrapping the compiled program into a ui.ImageFilter implementation.

For example:

ui.FragmentProgram currently has FragmentProgram.shader(...) here:
https://github.com/flutter/engine/blob/31b1a1d05803ec9fa25ae0f7b2b0f078ff3980b1/lib/ui/painting.dart#L3847

That method returns a ui.Shader object that you can use wherever flutter takes one of those objects (Paint.shader, etc.)

If ui.FragmentProgram had another method that returned a ui.ImageFilter object then you could use that wherever Flutter takes an ImageFilter (such as Paint.imageFilter, or the ImageFiltered widget, etc.)

**Comment by blasten** [comment]
> What would be needed is to allow the program to access an image input that will be supplied at render time 

I'm imagining that there be another scene builder operation, right? What would the API look like?

**Comment by blasten** [comment]
This will help implement https://github.com/flutter/flutter/issues/82906

**Comment by flar** [comment]
> > What would be needed is to allow the program to access an image input that will be supplied at render time
> 
> I'm imagining that there be another scene builder operation, right? What would the API look like?

ui.FragmentProgram currently has `FragmentProgram.shader(...)` here:
https://github.com/flutter/engine/blob/31b1a1d05803ec9fa25ae0f7b2b0f078ff3980b1/lib/ui/painting.dart#L3847

That method returns a ui.Shader object that you can use wherever flutter takes one of those objects (Paint.shader, etc.)

If ui.FragmentProgram had another method that returned a ui.ImageFilter object then you could use that wherever Flutter takes an ImageFilter (such as Paint.imageFilter, or the ImageFiltered widget, etc.)

I'll update the description with this information.

**Comment by antholeole** [comment]
This would be a helpful alternative solution to [the loupe](https://docs.google.com/document/d/1RWq0Kd_xsHR8j7IXtgc25feV3RV8Hzq2KSQEVaDDXUs/edit?resourcekey=0-qLXzpPdvKRwQM8nlCijclQ). It's currently working with BDF, but there are some side effects that make it feel slightly awkward to use. For instance, we can't style the loupe in the regular way (wrap it with a box shadow etc) since that shadow would show up in the loupe itself.

**Comment by flar** [comment]
> This would be a helpful alternative solution to [the loupe](https://docs.google.com/document/d/1RWq0Kd_xsHR8j7IXtgc25feV3RV8Hzq2KSQEVaDDXUs/edit?resourcekey=0-qLXzpPdvKRwQM8nlCijclQ). It's currently working with BDF, but there are some side effects that make it feel slightly awkward to use. For instance, we can't style the loupe in the regular way (wrap it with a box shadow etc) since that shadow would show up in the loupe itself.

How would allowing a custom ImageFilter help with the loupe issue? You'd still need BDF to apply it to the backdrop, wouldn't you?

**Comment by antholeole** [comment]
Right, it wouldn't solve the dependency on BDF - I misunderstood the solution. It would, though (I believe), allow us to replicate the fish eye ([reference](https://i.stack.imgur.com/EuFFJ.jpg) of what a fish eye effect is, since it's new to me) that [native android has](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/widget/Magnifier.java#81). I'd need to distort the image in the magnifier.

We'd concluded that the effect is so minor in the loupe that it isn't a blocker, but still listing the potential use case here for posterity.

**Comment by flar** [comment]
> This would be a helpful alternative solution to [the loupe](https://docs.google.com/document/d/1RWq0Kd_xsHR8j7IXtgc25feV3RV8Hzq2KSQEVaDDXUs/edit?resourcekey=0-qLXzpPdvKRwQM8nlCijclQ). It's currently working with BDF, but there are some side effects that make it feel slightly awkward to use. For instance, we can't style the loupe in the regular way (wrap it with a box shadow etc) since that shadow would show up in the loupe itself.

Could you stack the BDF with an empty container that has an outside-only shadow, but is drawn above (i.e. after) it?

**Comment by antholeole** [comment]
> > This would be a helpful alternative solution to [the loupe](https://docs.google.com/document/d/1RWq0Kd_xsHR8j7IXtgc25feV3RV8Hzq2KSQEVaDDXUs/edit?resourcekey=0-qLXzpPdvKRwQM8nlCijclQ). It's currently working with BDF, but there are some side effects that make it feel slightly awkward to use. For instance, we can't style the loupe in the regular way (wrap it with a box shadow etc) since that shadow would show up in the loupe itself.
> 
> Could you stack the BDF with an empty container that has an outside-only shadow, but is drawn above (i.e. after) it?

This is very similar to what ended up landing: https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/magnifier.dart#L402-L409   

The reason a clip is used is so that we don't have to tell the user "you shadow has to be outside only" they can pass in any shadow and it just works. The second reason is outside only looks very strange with an offset.

The 'slight awkwardness' I was talking about here (which is kind of unrelated to this issue because you're right that this landing won't solve it) is:

currently a user has to decorate a Magnifier with

```dart
Magnifier(
   decoration: Decoration(..),
)
```

and then we take the decoration and pass it in to the second child of the stack https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/magnifier.dart#L355-L364. This works fine (no complains so far...) 

But if a user does the following, they'll start to see unexpected results:

```dart
DecoratedBox(
  decoration: Decoration(..),
  child: Magnifier(...)
)
```

Again, though, this is not really related to this issue, or any issue at all, really. It's the only sensible way for BDF to work! I think the API is polished enough to the point that it's really not a problem.

**Comment by slightfoot** [comment]
If we had this feature we could really do these kinds of effects,
https://twitter.com/marcus_eckert/status/1569697533873602560

**Comment by v21** [comment]
I've been trying to figure out how to apply a blur with a BackdropFilter, but with soft edges to the blur (there are [a few](https://stackoverflow.com/questions/51526681/flutter-how-to-get-a-semitransparent-blurring-layer-with-a-hole-with-soft-edge) [other cases](https://stackoverflow.com/questions/43550853/how-do-i-do-the-frosted-glass-effect-in-flutter#comment106205020_43552634) online of people trying to achieve this). My conclusion is that this is currently not possible without modifications to the engine layer - but this issue would unblock that use case, even if a little awkwardly (both the blur effect & the soft mask would both have to be defined in shader code).

**Comment by flar** [comment]
> I've been trying to figure out how to apply a blur with a BackdropFilter, but with soft edges to the blur (there are [a few](https://stackoverflow.com/questions/51526681/flutter-how-to-get-a-semitransparent-blurring-layer-with-a-hole-with-soft-edge) [other cases](https://stackoverflow.com/questions/43550853/how-do-i-do-the-frosted-glass-effect-in-flutter#comment106205020_43552634) online of people trying to achieve this). My conclusion is that this is currently not possible without modifications to the engine layer - but this issue would unblock that use case, even if a little awkwardly (both the blur effect & the soft mask would both have to be defined in shader code).

Do you have a picture of the desired effect? Note that you could compose a filter from an existing blur and a custom soft-mask effect using ImageFilter.compose.

**Comment by v21** [comment]
Sure, so I'm making a game making tool focused on narrative games, and this is something I want to expose to the user. So an example of the effect plain, just to show the blur & the soft edge:

![photo-1477959858617-67f85cf4f1df](https://user-images.githubusercontent.com/136482/224394112-771a1da5-34c5-4535-957e-2369096fe241.jpeg)

and in the context of a game a user might make:
![8YGJao](https://user-images.githubusercontent.com/136482/224399949-8cd36cd1-49ca-4052-8c54-b54ad040998f.png)

Note that in this, the sprite and the backdrop both want to be blurred in this particular region, with the text drawn on top.

> compose a filter from an existing blur and a custom soft-mask effect using ImageFilter.compose

I'm not sure how I'd go about making the soft-mask part of this, without, well, shader support - I'd assumed that composing two blurs would just blur a bit extra rather than altering the area the blur is applied to. But I haven't actually tried it!

**Comment by cyberpwnn** [comment]
I'd love to try and see if a lower resolution blur would actually measurably improve the performance of blurring in flutter. Right now its pretty hard to stay in budget for an iphone at 120fps for example when you start using blurring all over the place. Perhaps a frag shader set for like 1/4th resolution might help this

**Comment by nikitadol** [comment]
@cyberpwnn https://github.com/flutter/flutter/issues/102211

**Comment by jonahwilliams** [comment]
This was done in https://github.com/flutter/engine/pull/53490 and will be in 3.29. Closing as fixed.
